leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```diff
@@ -178,14 +178,21 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
         try {
             return method.invoke(delegate, args);
         } catch (InvocationTargetException e) {
-            if (e.getCause() instanceof ServiceNotAvailableException
-                    || e.getCause() instanceof NotCurrentLeaderException) {
+            if (e.getTargetException() instanceof ServiceNotAvailableException
+                    || e.getTargetException() instanceof NotCurrentLeaderException) {
                 markAsNotLeading(leadershipToken, e.getCause());
             }
-            throw e.getCause();
+            if (e.getTargetException() instanceof InterruptedException && !(checkIfLeading(leadershipToken))) {
+                throw notCurrentLeaderException("received an interrupt due to leader election.", e.getTargetException());
+            }
+            throw e.getTargetException();
         }
     }
 
+    private boolean checkIfLeading(LeadershipToken leadershipToken) {
+        return leaderElectionService.isStillLeading(leadershipToken) == StillLeadingStatus.LEADING;
```

Comment:
I think this will result in a network call - we could alternatively just quickly check if the given token is equal to the current `leadershipTokenRef` (since it gets nulled out when we lose leadership)